TST: time/tests/test_precision.py failed in pyinstaller (computed error is different depending on the order of the arguments)

First failing log (2022-03-13): https://github.com/astropy/astropy/runs/5525474634

Last successful log (2022-03-12): https://github.com/astropy/astropy/runs/5519547613

Looks like this test was added in #10373 . Any idea how to fix, @Zac-HD or @mhvk ? 🙏 

https://github.com/astropy/astropy/blob/c7b0e928e82dc7a4e099124d5223700e5bb4cfe2/astropy/time/tests/test_precision.py#L313-L315

```
____________________________ test_two_sum_symmetric ____________________________

    @given(floats(), floats())
>   def test_two_sum_symmetric(f1, f2):

astropy_tests/time/tests/test_precision.py:314: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

f1 = -3.089785075544792e+307, f2 = 1.7976931348623157e+308

    @given(floats(), floats())
    def test_two_sum_symmetric(f1, f2):
>       np.testing.assert_equal(two_sum(f1, f2), two_sum(f2, f1))
E       AssertionError: 
E       Items are not equal:
E       item=1
E       
E        ACTUAL: nan
E        DESIRED: -9.9792015476736e+291

astropy_tests/time/tests/test_precision.py:315: AssertionError
----------------------------- Captured stdout call -----------------------------
Falsifying example: test_two_sum_symmetric(
    f1=-3.089785075544792e+307, f2=1.7976931348623157e+308,
)
```

--- a/astropy/time/tests/test_basic.py
+++ b/astropy/time/tests/test_basic.py
@@ -2399,7 +2399,7 @@ def test_linspace():
     """
     t1 = Time(['2021-01-01 00:00:00', '2021-01-02 00:00:00'])
     t2 = Time(['2021-01-01 01:00:00', '2021-12-28 00:00:00'])
-    atol = 1 * u.ps
+    atol = 2 * np.finfo(float).eps * abs(t1 - t2).max()
 
     ts = np.linspace(t1[0], t2[0], 3)
     assert ts[0].isclose(Time('2021-01-01 00:00:00'), atol=atol)
@@ -2409,13 +2409,13 @@ def test_linspace():
     ts = np.linspace(t1, t2[0], 2, endpoint=False)
     assert ts.shape == (2, 2)
     assert all(ts[0].isclose(Time(['2021-01-01 00:00:00', '2021-01-02 00:00:00']), atol=atol))
-    assert all(ts[1].isclose(Time(['2021-01-01 00:30:00', '2021-01-01 12:30:00']), atol=atol*10))
+    assert all(ts[1].isclose(Time(['2021-01-01 00:30:00', '2021-01-01 12:30:00']), atol=atol))
 
     ts = np.linspace(t1, t2, 7)
     assert ts.shape == (7, 2)
     assert all(ts[0].isclose(Time(['2021-01-01 00:00:00', '2021-01-02 00:00:00']), atol=atol))
-    assert all(ts[1].isclose(Time(['2021-01-01 00:10:00', '2021-03-03 00:00:00']), atol=atol*300))
-    assert all(ts[5].isclose(Time(['2021-01-01 00:50:00', '2021-10-29 00:00:00']), atol=atol*3000))
+    assert all(ts[1].isclose(Time(['2021-01-01 00:10:00', '2021-03-03 00:00:00']), atol=atol))
+    assert all(ts[5].isclose(Time(['2021-01-01 00:50:00', '2021-10-29 00:00:00']), atol=atol))
     assert all(ts[6].isclose(Time(['2021-01-01 01:00:00', '2021-12-28 00:00:00']), atol=atol))
 
 
@@ -2424,7 +2424,7 @@ def test_linspace_steps():
     """
     t1 = Time(['2021-01-01 00:00:00', '2021-01-01 12:00:00'])
     t2 = Time('2021-01-02 00:00:00')
-    atol = 1 * u.ps
+    atol = 2 * np.finfo(float).eps * abs(t1 - t2).max()
 
     ts, st = np.linspace(t1, t2, 7, retstep=True)
     assert ts.shape == (7, 2)
@@ -2441,7 +2441,7 @@ def test_linspace_fmts():
     t1 = Time(['2020-01-01 00:00:00', '2020-01-02 00:00:00'])
     t2 = Time(2458850, format='jd')
     t3 = Time(1578009600, format='unix')
-    atol = 1 * u.ps
+    atol = 2 * np.finfo(float).eps * abs(t1 - Time([t2, t3])).max()
 
     ts = np.linspace(t1, t2, 3)
     assert ts.shape == (3, 2)

--- a/astropy/time/tests/test_precision.py
+++ b/astropy/time/tests/test_precision.py
@@ -359,7 +359,7 @@ def test_day_frac_idempotent(i, f):
     assert (i_d, f_d) == day_frac(i_d, f_d)
 
 
-@given(integers(-2**52+2, 2**52-2), floats(-1, 1))
+@given(integers(-2**52+2, 2**52-int(erfa.DJM0)-3), floats(-1, 1))
 @example(i=65536, f=3.637978807091714e-12)
 def test_mjd_initialization_precise(i, f):
     t = Time(val=i, val2=f, format="mjd", scale="tai")
@@ -383,12 +383,12 @@ def test_day_frac_round_to_even(jd1, jd2):
     assert (abs(t_jd2) == 0.5) and (t_jd1 % 2 == 0)
 
 
-@given(scale=sampled_from(STANDARD_TIME_SCALES), jds=unreasonable_jd())
+@given(scale=sampled_from([sc for sc in STANDARD_TIME_SCALES if sc != 'utc']),
+       jds=unreasonable_jd())
 @example(scale="tai", jds=(0.0, 0.0))
 @example(scale="tai", jds=(0.0, -31738.500000000346))
 def test_resolution_never_decreases(scale, jds):
     jd1, jd2 = jds
-    assume(not scale == 'utc' or 2440000 < jd1 + jd2 < 2460000)
     t = Time(jd1, jd2, format="jd", scale=scale)
     with quiet_erfa():
         assert t != t + dt_tiny
@@ -397,11 +397,18 @@ def test_resolution_never_decreases(scale, jds):
 @given(reasonable_jd())
 @example(jds=(2442777.5, 0.9999999999999999))
 def test_resolution_never_decreases_utc(jds):
-    """UTC is very unhappy with unreasonable times"""
+    """UTC is very unhappy with unreasonable times,
+
+    Unlike for the other timescales, in which addition is done
+    directly, here the time is transformed to TAI before addition, and
+    then back to UTC.  Hence, some rounding errors can occur and only
+    a change of 2*dt_tiny is guaranteed to give a different time.
+
+    """
     jd1, jd2 = jds
     t = Time(jd1, jd2, format="jd", scale="utc")
     with quiet_erfa():
-        assert t != t + dt_tiny
+        assert t != t + 2*dt_tiny
 
 
 @given(scale1=sampled_from(STANDARD_TIME_SCALES),
@@ -429,6 +436,8 @@ def test_conversion_preserves_jd1_jd2_invariant(iers_b, scale1, scale2, jds):
        scale2=sampled_from(STANDARD_TIME_SCALES),
        jds=unreasonable_jd())
 @example(scale1='tai', scale2='utc', jds=(0.0, 0.0))
+@example(scale1='utc', scale2='ut1', jds=(2441316.5, 0.9999999999999991))
+@example(scale1='ut1', scale2='tai', jds=(2441498.5, 0.9999999999999999))
 def test_conversion_never_loses_precision(iers_b, scale1, scale2, jds):
     """Check that time ordering remains if we convert to another scale.
 
@@ -447,7 +456,9 @@ def test_conversion_never_loses_precision(iers_b, scale1, scale2, jds):
     try:
         with quiet_erfa():
             t2 = t + tiny
-            assert getattr(t, scale2) < getattr(t2, scale2)
+            t_scale2 = getattr(t, scale2)
+            t2_scale2 = getattr(t2, scale2)
+            assert t_scale2 < t2_scale2
     except iers.IERSRangeError:  # UT1 conversion needs IERS data
         assume(scale1 != 'ut1' or 2440000 < jd1 + jd2 < 2458000)
         assume(scale2 != 'ut1' or 2440000 < jd1 + jd2 < 2458000)
@@ -460,6 +471,19 @@ def test_conversion_never_loses_precision(iers_b, scale1, scale2, jds):
         barycentric = {scale1, scale2}.issubset({'tcb', 'tdb'})
         geocentric = {scale1, scale2}.issubset({'tai', 'tt', 'tcg'})
         assume(jd1 + jd2 >= -31738.5 or geocentric or barycentric)
+        raise
+    except AssertionError:
+        # Before 1972, TAI-UTC changed smoothly but not always very
+        # consistently; this can cause trouble on day boundaries for UTC to
+        # UT1; it is not clear whether this will ever be resolved (and is
+        # unlikely ever to matter).
+        # Furthermore, exactly at leap-second boundaries, it is possible to
+        # get the wrong leap-second correction due to rounding errors.
+        # The latter is xfail'd for now, but should be fixed; see gh-13517.
+        if 'ut1' in (scale1, scale2):
+            if abs(t_scale2 - t2_scale2 - 1 * u.s) < 1*u.ms:
+                pytest.xfail()
+            assume(t.jd > 2441317.5 or t.jd2 < 0.4999999)
         raise
 
 
@@ -484,9 +508,11 @@ def test_leap_stretch_mjd(d, f):
          jds=(2441682.5, 2.2204460492503136e-16),
          delta=7.327471962526035e-12)
 @example(scale='utc', jds=(0.0, 5.787592627370942e-13), delta=0.0)
+@example(scale='utc', jds=(1.0, 0.25000000023283064), delta=-1.0)
 def test_jd_add_subtract_round_trip(scale, jds, delta):
     jd1, jd2 = jds
-    if scale == 'utc' and abs(jd1+jd2) < 1:
+    if scale == 'utc' and (jd1+jd2 < 1
+                           or jd1+jd2+delta < 1):
         # Near-zero UTC JDs degrade accuracy; not clear why,
         # but also not so relevant, so ignoring.
         thresh = 100*u.us
@@ -505,20 +531,25 @@ def test_jd_add_subtract_round_trip(scale, jds, delta):
         raise
 
 
-@given(scale=sampled_from(STANDARD_TIME_SCALES),
+@given(scale=sampled_from(TimeDelta.SCALES),
        jds=reasonable_jd(),
        delta=floats(-3*tiny, 3*tiny))
 @example(scale='tai', jds=(0.0, 3.5762786865234384), delta=2.220446049250313e-16)
 @example(scale='tai', jds=(2441316.5, 0.0), delta=6.938893903907228e-17)
 @example(scale='tai', jds=(2441317.5, 0.0), delta=-6.938893903907228e-17)
-@example(scale='tai', jds=(2441317.5, 0.0), delta=-6.938893903907228e-17)
+@example(scale='tai', jds=(2440001.0, 0.49999999999999994), delta=5.551115123125783e-17)
 def test_time_argminmaxsort(scale, jds, delta):
     jd1, jd2 = jds
-    t = Time(jd1, jd2+np.array([0, delta]), scale=scale, format="jd")
+    t = (Time(jd1, jd2, scale=scale, format="jd")
+         + TimeDelta([0, delta], scale=scale, format='jd'))
     imin = t.argmin()
     imax = t.argmax()
     isort = t.argsort()
-    diff = (t.jd1[1]-t.jd1[0]) + (t.jd2[1]-t.jd2[0])
+    # Be careful in constructing diff, for case that abs(jd2[1]-jd2[0]) ~ 1.
+    # and that is compensated by jd1[1]-jd1[0] (see example above).
+    diff, extra = two_sum(t.jd2[1], -t.jd2[0])
+    diff += t.jd1[1]-t.jd1[0]
+    diff += extra
     if diff < 0:  # item 1 smaller
         assert delta < 0
         assert imin == 1 and imax == 0 and np.all(isort == [1, 0])

--- a/astropy/time/tests/test_precision.py
+++ b/astropy/time/tests/test_precision.py
@@ -310,10 +310,12 @@ def test_two_sum(i, f):
         assert_almost_equal(a, b, atol=Decimal(tiny), rtol=Decimal(0))
 
 
-# For why bounds are set, see
-# https://github.com/astropy/astropy/issues/12955#issuecomment-1186293703
-@given(floats(min_value=np.finfo('f8').min * 0.5, max_value=np.finto('f8').max * 0.5),
-       floats(min_value=np.finfo('f8').min * 0.5, max_value=np.finto('f8').max * 0.5))
+# The bounds are here since we want to be sure the sum does not go to infinity,
+# which does not have to be completely symmetric; e.g., this used to fail:
+#     @example(f1=-3.089785075544792e307, f2=1.7976931348623157e308)
+# See https://github.com/astropy/astropy/issues/12955#issuecomment-1186293703
+@given(floats(min_value=np.finfo(float).min/2, max_value=np.finfo(float).max/2),
+       floats(min_value=np.finfo(float).min/2, max_value=np.finfo(float).max/2))
 def test_two_sum_symmetric(f1, f2):
     np.testing.assert_equal(two_sum(f1, f2), two_sum(f2, f1))
 
@@ -346,8 +348,6 @@ def test_day_frac_harmless(i, f):
 def test_day_frac_exact(i, f):
     assume(abs(f) < 0.5 or i % 2 == 0)
     i_d, f_d = day_frac(i, f)
-    # FIXME: Might be a real bug we need to fix in time/utils.py
-    # i = 1, f = 0.49999999999999994 --> (2, -0.5)
     assert i == i_d
     assert f == f_d
 

--- a/astropy/time/tests/test_precision.py
+++ b/astropy/time/tests/test_precision.py
@@ -310,7 +310,10 @@ def test_two_sum(i, f):
         assert_almost_equal(a, b, atol=Decimal(tiny), rtol=Decimal(0))
 
 
-@given(floats(), floats())
+# For why bounds are set, see
+# https://github.com/astropy/astropy/issues/12955#issuecomment-1186293703
+@given(floats(min_value=np.finfo('f8').min * 0.5, max_value=np.finto('f8').max * 0.5),
+       floats(min_value=np.finfo('f8').min * 0.5, max_value=np.finto('f8').max * 0.5))
 def test_two_sum_symmetric(f1, f2):
     np.testing.assert_equal(two_sum(f1, f2), two_sum(f2, f1))
 
@@ -339,9 +342,12 @@ def test_day_frac_harmless(i, f):
 
 @given(integers(-2**52+2, 2**52-2), floats(-0.5, 0.5))
 @example(i=65536, f=3.637978807091714e-12)
+@example(i=1, f=0.49999999999999994)
 def test_day_frac_exact(i, f):
     assume(abs(f) < 0.5 or i % 2 == 0)
     i_d, f_d = day_frac(i, f)
+    # FIXME: Might be a real bug we need to fix in time/utils.py
+    # i = 1, f = 0.49999999999999994 --> (2, -0.5)
     assert i == i_d
     assert f == f_d
 
@@ -389,6 +395,7 @@ def test_resolution_never_decreases(scale, jds):
 
 
 @given(reasonable_jd())
+@example(jds=(2442777.5, 0.9999999999999999))
 def test_resolution_never_decreases_utc(jds):
     """UTC is very unhappy with unreasonable times"""
     jd1, jd2 = jds
@@ -502,6 +509,9 @@ def test_jd_add_subtract_round_trip(scale, jds, delta):
        jds=reasonable_jd(),
        delta=floats(-3*tiny, 3*tiny))
 @example(scale='tai', jds=(0.0, 3.5762786865234384), delta=2.220446049250313e-16)
+@example(scale='tai', jds=(2441316.5, 0.0), delta=6.938893903907228e-17)
+@example(scale='tai', jds=(2441317.5, 0.0), delta=-6.938893903907228e-17)
+@example(scale='tai', jds=(2441317.5, 0.0), delta=-6.938893903907228e-17)
 def test_time_argminmaxsort(scale, jds, delta):
     jd1, jd2 = jds
     t = Time(jd1, jd2+np.array([0, delta]), scale=scale, format="jd")

--- a/astropy/time/utils.py
+++ b/astropy/time/utils.py
@@ -60,14 +60,16 @@ def day_frac(val1, val2, factor=None, divisor=None):
 
     # get integer fraction
     day = np.round(sum12)
-    extra, frac = two_sum(sum12, -day)
-    frac += extra + err12
-    # Our fraction can now have gotten >0.5 or <-0.5, which means we would
-    # loose one bit of precision. So, correct for that.
-    excess = np.round(frac)
+    # Calculate remaining fraction. This can have gotten >0.5 or <-0.5, which means
+    # we would lose one bit of precision. So, correct for that.  Here, we need
+    # particular care for the case that frac=0.5 and check>0 or frac=-0.5 and check<0,
+    # since in that case if check is large enough, rounding was done the wrong way.
+    frac, check = two_sum(sum12 - day, err12)
+    excess = np.where(frac * np.sign(check) != 0.5, np.round(frac),
+                      np.round(frac+2*check))
     day += excess
-    extra, frac = two_sum(sum12, -day)
-    frac += extra + err12
+    frac = sum12 - day
+    frac += err12
     return day, frac
 
 

--- a/setup.cfg
+++ b/setup.cfg
@@ -55,14 +55,12 @@ asdf_extensions =
 [options.extras_require]
 test =  # Required to run the astropy test suite.
     pytest>=7.0
-    hypothesis==6.46.7
     pytest-doctestplus>=0.12
     pytest-astropy-header>=0.2.1
     pytest-astropy>=0.10
     pytest-xdist
 test_all =  # Required for testing, plus packages used by particular tests.
     pytest>=7.0
-    hypothesis==6.46.7
     pytest-doctestplus>=0.12
     pytest-astropy-header>=0.2.1
     pytest-astropy>=0.10

